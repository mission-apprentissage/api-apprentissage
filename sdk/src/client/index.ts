export * from "./generateUri/generateUri.js";

export * from "./parser/response.parser.js";

export * from "./apiError.js";

export * from "./client.js";
