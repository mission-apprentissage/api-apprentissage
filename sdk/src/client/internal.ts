export * from "./internal/generateUri/generateUri.js";

export * from "./internal/parser/response.parser.js";

export * from "./internal/apiError.js";
