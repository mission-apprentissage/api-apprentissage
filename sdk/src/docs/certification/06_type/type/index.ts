import { DocField } from "../../../types.js";
import description from "./description.md.js";
import information from "./information.md.js";

export const typeField: DocField = {
  name: "type",
  description,
  information,
  sample: null,
  tags: [
    ".nature.cfd.code",
    ".nature.cfd.libelle",
    ".gestionnaire_diplome",
    ".certificateurs_rncp[].siret",
    ".certificateurs_rncp[].nom",
    ".enregistrement_rncp",
    ".voie_acces.rncp",
  ],
};
