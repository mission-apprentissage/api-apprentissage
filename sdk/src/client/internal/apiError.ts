import type { IResErrorJson } from "../../models/errors/errors.model.js";
import type { PathParam, QueryString, WithQueryStringAndPathParam } from "./generateUri/generateUri.js";

export interface ApiErrorContext {
  path: string;
  params: PathParam;
  querystring: QueryString;
  requestHeaders: Record<string, string | string[]>;
  statusCode: number;
  message: string;
  name: string;
  responseHeaders: Record<string, string | string[]>;
  errorData: unknown;
}

export class ApiError extends Error {
  context: ApiErrorContext;

  constructor(context: ApiErrorContext) {
    super();
    this.context = context;
    this.name = context.name ?? "ApiError";
    this.message = context.message ?? `code ${context.statusCode}`;
  }

  toJSON(): ApiErrorContext {
    return this.context;
  }

  static async build(
    path: string,
    requestHeaders: Readonly<Headers>,
    options: WithQueryStringAndPathParam,
    res: Response
  ): Promise<ApiError> {
    let message = res.status === 0 ? "Network Error" : res.statusText;
    let name = "Api Error";
    let errorData: IResErrorJson["data"] | null = null;

    if (res.status > 0) {
      try {
        if (res.headers.get("Content-Type")?.startsWith("application/json")) {
          const data = await res.json();
          if (typeof data === "object" && data !== null) {
            name = "name" in data && typeof data.name === "string" ? data.name : "Api Error";
            message = "message" in data && typeof data.message === "string" ? data.message : `code ${res.status}`;
            errorData = "data" in data ? data.data : null;
          }
        }
      } catch (_error) {
        // ignore
      }
    }

    return new ApiError({
      path,
      params: "params" in options && options.params ? options.params : {},
      querystring: "querystring" in options && options.querystring ? options.querystring : {},
      requestHeaders: Object.fromEntries(requestHeaders.entries()),
      statusCode: res.status,
      message,
      name,
      responseHeaders: Object.fromEntries(res.headers.entries()),
      errorData,
    });
  }
}
