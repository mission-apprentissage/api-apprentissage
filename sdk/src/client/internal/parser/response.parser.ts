import { z } from "zod/v4-mini";
import { $ZodError, treeifyError } from "zod/v4/core";
import type { $ZodType } from "zod/v4/core";

export class ApiParseError extends Error {
  constructor(zodError: $ZodError) {
    super(
      "api-alternance-sdk: Error while parsing API response, this is either a bug in the SDK or a change in the API response. Please report this issue.\n" +
        JSON.stringify(treeifyError(zodError))
    );
    this.name = "ApiParseError";
    this.cause = zodError;
  }
}

export function parseApiResponse<T extends $ZodType>(data: unknown, schema: T): z.output<T> {
  try {
    return z.parse(schema, data);
  } catch (error) {
    if (error instanceof $ZodError) {
      throw new ApiParseError(error);
    }
    throw error;
  }
}
