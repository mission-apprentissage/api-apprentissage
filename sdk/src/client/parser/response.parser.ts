import type { z, ZodType } from "zod";
import { ZodError } from "zod";

export class ApiParseError extends Error {
  constructor(zodError: ZodError) {
    super(
      "api-alternance-sdk: Error while parsing API response, this is either a bug in the SDK or a change in the API response. Please report this issue.\n\tZodErrors: " +
        zodError.toString().split("\n").join("\n\t")
    );
    this.name = "ApiParseError";
    this.cause = zodError;
  }
}

export function parseApiResponse<T extends ZodType>(data: unknown, schema: T): z.output<T> {
  try {
    return schema.parse(data);
  } catch (error) {
    if (error instanceof ZodError) {
      throw new ApiParseError(error);
    }
    throw error;
  }
}
