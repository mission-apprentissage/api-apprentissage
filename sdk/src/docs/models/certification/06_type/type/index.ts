import type { DocBusinessField } from "../../../../types.js";
import description from "./description.md.js";

export default <DocBusinessField>{
  type: "business",
  name: "type",
  description,
  information: null,
  sample: null,
  notes: null,
  tip: null,
  tags: [
    ".nature.cfd.code",
    ".nature.cfd.libelle",
    ".gestionnaire_diplome",
    ".certificateurs_rncp[].siret",
    ".certificateurs_rncp[].nom",
    ".enregistrement_rncp",
    ".voie_acces.rncp",
  ],
};
