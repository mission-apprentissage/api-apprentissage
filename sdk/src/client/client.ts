import type { EmptyObject } from "type-fest";
import type { z, ZodType } from "zod";

import type { IApiRouteSchema, IApiRouteSchemaWrite } from "../routes/common.routes.js";
import type {
  IApiDeleteRoutes,
  IApiGetRoutes,
  IApiPostRoutes,
  IApiPutRoutes,
  IApiRequest,
  IApiResponse,
} from "../routes/index.js";
import { ApiError } from "./apiError.js";
import type { CertificationModule } from "./certification/certification.module.js";
import { buildCertificationModule } from "./certification/certification.module.js";
import type { WithQueryStringAndPathParam } from "./generateUri/generateUri.js";
import { generateUri } from "./generateUri/generateUri.js";
import type { JobModule } from "./job/job.module.js";
import { buildJobModule } from "./job/job.module.js";
import type { OrganismeModule } from "./organisme/organisme.module.js";
import { buildOrganismeModule } from "./organisme/organisme.module.js";

type OptionsGet = {
  [Prop in keyof Pick<IApiRouteSchema, "params" | "querystring" | "headers">]: IApiRouteSchema[Prop] extends ZodType
    ? z.input<IApiRouteSchema[Prop]>
    : never;
};

type OptionsWrite = {
  [Prop in keyof Pick<
    IApiRouteSchemaWrite,
    "params" | "querystring" | "headers" | "body"
  >]: IApiRouteSchemaWrite[Prop] extends ZodType ? z.input<IApiRouteSchemaWrite[Prop]> : never;
};

type IRequestOptions = OptionsGet | OptionsWrite | EmptyObject;

type FetchOptions = Omit<RequestInit, "body" | "method" | "headers">;

const removeAtEnd = (url: string, removed: string): string =>
  url.endsWith(removed) ? url.slice(0, -removed.length) : url;

function throwError(message: string): never {
  throw new Error(message);
}

export type ApiClientConfig = {
  endpoint?: string;
  key: string;
};

export class ApiClient {
  endpoint: string;
  key: string;

  certification: CertificationModule;
  job: JobModule;
  organisme: OrganismeModule;

  constructor(config: ApiClientConfig) {
    this.endpoint = removeAtEnd(config.endpoint ?? "https://api.apprentissage.beta.gouv.fr/api", "/");
    this.key = config.key ?? throwError("api-alternance-sdk: api key is required");

    this.certification = buildCertificationModule(this);
    this.job = buildJobModule(this);
    this.organisme = buildOrganismeModule(this);
  }

  private buildRequestInit(
    method: RequestInit["method"],
    options: IRequestOptions,
    rawOptions?: FetchOptions
  ): RequestInit {
    const headers = new Headers();
    headers.set("authorization", `Bearer ${this.key}`);

    if ("headers" in options && options.headers) {
      const h = options.headers;
      Object.keys(h).forEach((name) => {
        headers.set(name, h[name]);
      });
    }

    let body: RequestInit["body"] | undefined = undefined;
    if ("body" in options && method !== "GET") {
      // @ts-expect-error
      if (options.body instanceof FormData) {
        body = options.body;
      } else {
        body = JSON.stringify(options.body);
        headers.set("content-Type", "application/json");
      }
    }

    const requestInit: RequestInit = {
      body,
      method,
      headers,
    };

    if (rawOptions) {
      Object.assign(requestInit, rawOptions);
    }

    return requestInit;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private async exec(path: string, requestInit: RequestInit, options: WithQueryStringAndPathParam): Promise<any> {
    const reqHeaders = Array.from((requestInit.headers as Headers)?.entries() ?? []);
    const res = await fetch(this.endpoint + generateUri(path, options), requestInit);

    if (!res.ok) {
      throw await ApiError.build(path, new Headers(reqHeaders), options, res);
    }

    if (res.status === 204) {
      return;
    }

    return res.json();
  }

  async get<P extends keyof IApiGetRoutes, S extends IApiGetRoutes[P] = IApiGetRoutes[P]>(
    path: P,
    options: IApiRequest<S>,
    rawOptions?: FetchOptions
  ): Promise<IApiResponse<S>> {
    const requestInit = this.buildRequestInit("GET", options, rawOptions);
    return this.exec(path, requestInit, options);
  }

  async post<P extends keyof IApiPostRoutes, S extends IApiPostRoutes[P] = IApiPostRoutes[P]>(
    path: P,
    options: IApiRequest<S>,
    rawOptions?: FetchOptions
  ): Promise<IApiResponse<S>> {
    const requestInit = this.buildRequestInit("POST", options, rawOptions);
    return this.exec(path, requestInit, options);
  }

  async put<P extends keyof IApiPutRoutes, S extends IApiPutRoutes[P] = IApiPutRoutes[P]>(
    path: P,
    options: IApiRequest<S>,
    rawOptions?: FetchOptions
  ): Promise<IApiResponse<S>> {
    const requestInit = this.buildRequestInit("PUT", options, rawOptions);
    return this.exec(path, requestInit, options);
  }

  async delete<P extends keyof IApiDeleteRoutes, S extends IApiDeleteRoutes[P] = IApiDeleteRoutes[P]>(
    path: P,
    options: IApiRequest<S>,
    rawOptions?: FetchOptions
  ): Promise<IApiResponse<S>> {
    const requestInit = this.buildRequestInit("DELETE", options, rawOptions);
    return this.exec(path, requestInit, options);
  }
}
