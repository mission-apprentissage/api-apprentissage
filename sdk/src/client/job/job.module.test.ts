import nock, { cleanAll, disableNetConnect, enableNetConnect } from "nock";
import { beforeEach, describe, expect, expectTypeOf, it } from "vitest";

import type { IJobSearchQuery, IJobSearchResponse } from "../../routes/job.routes.js";
import { ApiError } from "../apiError.js";
import { ApiClient } from "../client.js";
import { ApiParseError } from "../parser/response.parser.js";
import type { JobSearchFilter } from "./job.module.js";

beforeEach(() => {
  disableNetConnect();

  return () => {
    cleanAll();
    enableNetConnect();
  };
});

describe("search", () => {
  const response: IJobSearchResponse = {
    jobs: [
      {
        identifier: {
          id: "1",
          partner: "La bonne alternance",
          partner_job_id: null,
        },
        workplace: {
          siret: "11000001500013",
          brand: "Brand",
          legal_name: "ASSEMBLEE NATIONALE",
          website: null,
          name: "ASSEMBLEE NATIONALE",
          description: "Workplace Description",
          size: null,
          address: {
            label: "Paris",
          },
          geopoint: {
            coordinates: [2.347, 48.8589],
            type: "Point",
          },
          idcc: 1242,
          opco: "",
          naf: {
            code: "84.11Z",
            label: "Autorité constitutionnelle",
          },
        },
        apply: {
          url: "https://postler.com",
          phone: "0300000000",
        },
        contract: {
          start: new Date("2021-01-28T16:00:00.000+01:00"),
          duration: 12,
          type: ["Apprentissage"],
          remote: "onsite",
        },
        offer: {
          title: "Opérations administratives",
          rome_codes: ["M1602"],
          description: "Exécute des travaux administratifs courants",
          target_diploma: {
            european: "4",
            label: "BP, Bac, autres formations niveau (Bac)",
          },
          desired_skills: ["Faire preuve de rigueur et de précision"],
          to_be_acquired_skills: [
            "Production, Fabrication: Procéder à l'enregistrement, au tri, à l'affranchissement du courrier",
            "Production, Fabrication: Réaliser des travaux de reprographie",
            "Organisation: Contrôler la conformité des données ou des documents",
          ],
          access_conditions: ["Ce métier est accessible avec un diplôme de fin d'études secondaires"],
          creation: new Date("2021-01-01T01:00:00.000+01:00"),
          expiration: new Date("2021-03-28T17:00:00.000+02:00"),
          opening_count: 1,
          status: "Active",
        },
      },
    ],
    recruiters: [
      {
        identifier: {
          id: "42",
        },
        workplace: {
          siret: "11000001500013",
          brand: "ASSEMBLEE NATIONALE - La vraie",
          legal_name: "ASSEMBLEE NATIONALE",
          website: null,
          name: "ASSEMBLEE NATIONALE - La vraie",
          description: null,
          size: null,
          address: {
            label: "126 RUE DE L'UNIVERSITE 75007 PARIS",
          },
          geopoint: {
            coordinates: [2.347, 48.8589],
            type: "Point",
          },
          idcc: null,
          opco: null,
          naf: {
            code: "8411Z",
            label: "Administration publique générale",
          },
        },
        apply: {
          url: "http://localhost:3000/recherche-apprentissage?type=lba&itemId=11000001500013",
          phone: "0100000000",
        },
      },
    ],
    warnings: [
      {
        code: "FRANCE_TRAVAIL_API_ERROR",
        message: "Unable to retrieve job offers from France Travail API",
      },
    ],
  };

  const query: Required<IJobSearchQuery> = {
    longitude: -4.6,
    latitude: 42.85,
    radius: 60,
    target_diploma_level: "3",
    romes: "I1401,I1306",
    rncp: "RNCP38654",
  };
  const filter: JobSearchFilter = {
    longitude: -4.6,
    latitude: 42.85,
    radius: 60,
    target_diploma_level: "3",
    romes: ["I1401", "I1306"],
    rncp: "RNCP38654",
  };

  it("should call the API with the correct querystring", async () => {
    const scope = nock("https://api.apprentissage.beta.gouv.fr/api", {
      reqheaders: { authorization: "Bearer api-key" },
    })
      .get("/job/v1/search")
      .query(query)
      .reply(200, response);

    const apiClient = new ApiClient({ key: "api-key" });

    const data = await apiClient.job.search(filter);

    expectTypeOf(data).toEqualTypeOf<IJobSearchResponse>();

    expect(scope.isDone()).toBe(true);
    expect(data).toEqual(response);
  });

  it("should throw an ApiError when server error", async () => {
    const scope = nock("https://api.apprentissage.beta.gouv.fr/api", {
      reqheaders: { authorization: "Bearer api-key" },
    })
      .get("/job/v1/search")
      .query(query)
      .reply(401, {
        statusCode: 401,
        name: "Unauthorized",
        message: "Vous devez fournir une clé d'API valide pour accéder à cette ressource",
      });

    const apiClient = new ApiClient({ key: "api-key" });
    const err = await apiClient.job
      .search(filter)
      .then(() => {
        expect.unreachable("should throw an error");
      })
      .catch((error: ApiError) => {
        return error;
      });

    expect(err).toBeInstanceOf(ApiError);
    expect(err.name).toBe("Unauthorized");

    expect(scope.isDone()).toBe(true);
  });

  it("should throw if the response does not match the schema", async () => {
    const scope = nock("https://api.apprentissage.beta.gouv.fr/api", {
      reqheaders: { authorization: "Bearer api-key" },
    })
      .get("/job/v1/search")
      .query(query)
      .reply(200, { breaking: "schema" });

    const apiClient = new ApiClient({ key: "api-key" });
    const err = await apiClient.job
      .search(filter)
      .then(() => {
        expect.unreachable("should throw an error");
      })
      .catch((error: ApiError) => {
        return error;
      });

    expect(err).toBeInstanceOf(ApiParseError);
    expect(err.name).toBe("ApiParseError");
    expect(err.message).toMatchSnapshot();

    expect(scope.isDone()).toBe(true);
  });

  it("should accepts future schema ehancements", async () => {
    const scope = nock("https://api.apprentissage.beta.gouv.fr/api", {
      reqheaders: { authorization: "Bearer api-key" },
    })
      .get("/job/v1/search")
      .query(query)
      .reply(200, {
        ...response,
        new_field: "new_field",
      });

    const apiClient = new ApiClient({ key: "api-key" });

    const data = await apiClient.job.search(filter);

    expectTypeOf(data).toEqualTypeOf<IJobSearchResponse>();

    expect(scope.isDone()).toBe(true);
    expect(data).toEqual(response);
  });
});
