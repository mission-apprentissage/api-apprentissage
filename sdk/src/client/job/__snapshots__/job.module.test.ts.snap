// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createOffer > should throw if the response does not match the schema 1`] = `
"api-alternance-sdk: Error while parsing API response, this is either a bug in the SDK or a change in the API response. Please report this issue.
	ZodErrors: [
	  {
	    "code": "invalid_type",
	    "expected": "string",
	    "received": "undefined",
	    "path": [
	      "id"
	    ],
	    "message": "Required"
	  }
	]"
`;

exports[`search > should throw if the response does not match the schema 1`] = `
"api-alternance-sdk: Error while parsing API response, this is either a bug in the SDK or a change in the API response. Please report this issue.
	ZodErrors: [
	  {
	    "code": "invalid_type",
	    "expected": "array",
	    "received": "undefined",
	    "path": [
	      "jobs"
	    ],
	    "message": "Required"
	  },
	  {
	    "code": "invalid_type",
	    "expected": "array",
	    "received": "undefined",
	    "path": [
	      "recruiters"
	    ],
	    "message": "Required"
	  },
	  {
	    "code": "invalid_type",
	    "expected": "array",
	    "received": "undefined",
	    "path": [
	      "warnings"
	    ],
	    "message": "Required"
	  }
	]"
`;
